template/classroom_setup.sh.erb

```diff
@@ -0,0 +1,128 @@
+      exit 1
+    fi
+  else
+    echo "Cannot find any project directory for {project_id}"
+    exit 1
+  fi
+}
+
+find_project_owner() {
+  project_dir=$(find_project_directory)
+  echo $(getfacl -t project_dir |& grep USER |awk '{print $2}')
+}
+
+create_classroom_directory() {
+  class_dir=$1
+
+  echo "Creating class directory $class_dir" 
+  umask 027
+  mkdir -v -p $class_dir
+  chgrp $project_id $class_dir
+  chmod g+s $class_dir
+  setfacl -R -d -m g:${project_id}:r-X $class_dir
+
+  echo "Creating materials directory $class_dir/materials"
+  mkdir -v -p $class_dir/materials
+
+  echo "Creating workspace directory $class_dir/workspace"
+  mkdir -v -p $class_dir/workspace
+}
+
+create_jupyter_classroom() {
+  project_dir="$1"
+  class_dir="$project_dir/$class_id"
+  jupyter_dir="$class_dir/jupyter"
+
+  if [[ -d $jupyter_dir ]]; then
+    echo "The Jupyter environment $jupyter_dir has been created"
+    return 0
+  fi
+
+  if [[ ! -d $class_dir ]]; then
+    create_classroom_directory $class_dir
+  fi
+
+  echo "Creating Jupyter environment with $(which python3)"
+  python3 -m venv --without-pip --prompt "$class_id" $jupyter_dir
+  source $jupyter_dir/bin/activate
+  # curl https://bootstrap.pypa.io/get-pip.py |python # RHEL9
+  curl https://bootstrap.pypa.io/pip/3.6/get-pip.py |python # RHEL7
+  pip install \
+	--log=${class_dir}/setup_jupyter.log --isolated \
+       	--no-cache-dir --ignore-installed \
+	jupyterlab notebook jupyter-console qtconsole ipywidgets
+  deactivate
+
+  echo "Setting up PDF exporter"
+  #pandoc_version=3.6.4 # RHEL9
+  pandoc_version=2.9.2  # RHEL7
+  pushd $jupyter_dir
+  wget -nc https://github.com/jgm/pandoc/releases/download/${pandoc_version}/pandoc-${pandoc_version}-linux-amd64.tar.gz
+  tar xf pandoc-${pandoc_version}-linux-amd64.tar.gz --strip=1
+  rm -f pandoc-${pandoc_version}-linux-amd64.tar.gz
+  test -d share/jupyter/nbconvert/templates/html || mkdir -p share/jupyter/nbconvert/templates/html
+  chmod -v g+rx $jupyter_dir/share/jupyter/nbconvert/templates/html
+  
+  echo "Disabling the group permission of Jupyter environment"
+  chmod g-w -R ${jupyter_dir}
+}
+
+project_dir=$(find_project_directory)
+project_owner=$(find_project_owner)
+
+# Only inspector (project owner) can create the classroom environment
+if [[ "$USER" == "$project_owner" ]]; then
+  create_jupyter_classroom $project_dir
+fi
+
+class_dir="$project_dir/$class_id"
+export OSC_CLASS_DIR="$class_dir"
+
+# Define material directory
+export OSC_CLASS_FILES="$class_dir/materials"
+
+# Load required modules
+LMOD_EXACT_MATCH=no module load texlive nodejs
+
+# Setup Jupyter
+export JUPYTER_DATA_DIR="${HOME}/osc_classes/${class_id}/.local/share/jupyter"
+jupyter_dir="$class_dir/jupyter"
+ls $jupyter_dir > /dev/null 2>&1
+if [[ $? == 0 ]]; then 
+  source ${jupyter_dir}/bin/activate
+else
+  echo "Cannot access ${jupyter_dir}. Please contact OSC help or class instructor." >&2
+fi
+
+export PATH="/users/PZS0645/support/classroom/bin:$PATH"
```

Comment:
Do we need this? And/or should it be a script here in this repository?

Reply:
We can remove it for now. I added a script for a specific classroom that allows students to submit their notebooks to run on a regular GPU node, because the classroom’s GPU node does not have enough resources.